src/Middleware/HttpLogging/src/BufferingStream.cs

```diff
@@ -0,0 +1,303 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.Diagnostics;
+using System.IO;
+using System.IO.Pipelines;
+using System.Runtime.CompilerServices;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNetCore.HttpLogging
+{
+    internal class BufferingStream : Stream, IBufferWriter<byte>
+    {
+        private const int MinimumBufferSize = 4096; // 4K
+        protected int _bytesBuffered;
+        private BufferSegment? _head;
+        private BufferSegment? _tail;
+        protected Memory<byte> _tailMemory; // remainder of tail memory
+        protected int _tailBytesBuffered;
+        protected ILogger _logger;
+        protected Stream _innerStream;
+
+        public BufferingStream(Stream innerStream, ILogger logger)
+        {
+            _logger = logger;
+            _innerStream = innerStream;
+        }
+
+        public override bool CanSeek => _innerStream.CanSeek;
+
+        public override bool CanRead => _innerStream.CanRead;
+
+        public override bool CanWrite => _innerStream.CanWrite;
+
+        public override long Length => _innerStream.Length;
+
+        public override long Position
+        {
+            get => _innerStream.Position;
+            set => _innerStream.Position = value;
+        }
+
+        public override int WriteTimeout
+        {
+            get => _innerStream.WriteTimeout;
+            set => _innerStream.WriteTimeout = value;
+        }
+
+        public string GetString(Encoding? encoding)
+        {
+            try
+            {
+                if (_head == null || _tail == null)
+                {
+                    // nothing written
+                    return "";
+                }
+
+                if (encoding == null)
+                {
+                    _logger.UnrecognizedMediaType();
+                    return "<Unrecognized media type>";
+                }
+
+                // Only place where we are actually using the buffered data.
+                // update tail here.
+                _tail.End = _tailBytesBuffered;
+
+                var ros = new ReadOnlySequence<byte>(_head, 0, _tail, _tailBytesBuffered);
+
+                var bufferWriter = new ArrayBufferWriter<char>();
+
+                var decoder = encoding.GetDecoder();
+                EncodingExtensions.Convert(decoder, ros, bufferWriter, flush: false, out var charUsed, out var completed);
+
+                while (true)
```

Comment:
What's the loop for?

Are there test(s) for sliced multi-byte payloads?

Reply:
Advice from @GrabYourPitchforks .  He recommended calling Convert in a loop after the first call with an empty ROS until completed is true if the first call to convert had completed `false`.

> Are there test(s) for sliced multi-byte payloads?

https://github.com/dotnet/aspnetcore/pull/31816/files#diff-b1f8d7f82e5eb842546da37a6f8471632e34d65762b26172f56a5eabccae2f2dR359

Reply:
Looks like it'll throw an `ArgumentException`? https://source.dot.net/#System.Private.CoreLib/Encoder.cs,320
That wont be caught below.

Reply:
No, this does hit a boundary. Each character is 3 bytes and the limit is 32Kb. It hits the boundary. I could do a smaller test, but I believe this to be sufficient.

I didn't personally look into this API directly, just the advice from Levi here. I'm curious if we need to call this in a loop at all, and just call it once and log the result in the ArrayWriter.

Should I just add a try catch for any exception type here?

Reply:
Idk, but the test currently doesn't hit line 89, so something is wrong with it.

Reply:
I also don't think it'll throw that exception at all because we always provide a large enough output buffer. https://docs.microsoft.com/en-us/dotnet/api/system.text.encoder.convert?view=net-5.0

>ArgumentException
The output buffer is too small to contain any of the converted input. The output buffer should be greater than or equal to the size indicated by the GetByteCount method.

Reply:
> Idk, but the test currently doesn't hit line 89, so something is wrong with it.

I noticed that, the first call to it returns `completed: true`. The input hits the boundary though. Hence I was asking earlier

> I'm curious if we need to call this in a loop at all, and just call it once and log the result in the ArrayWriter.

Reply:
This is what Levi told me:
> This is expected. flush: true means "I am never going to send you any more data. The byte sequence is complete."
completed: true [as a return value] means "I have processed the entire input buffer you gave me and cannot make any more forward progress."

> It's possible to get completed: true for flush: false. That's the decoder telling you "I need you to hand me more data before I can give you any more chars."

> Another way to think about it: flush=true means "assume there's <EOF> after this buffer."

@GrabYourPitchforks do you have an example of a byte sequence that would cause the first call to Convert to return completed false, and then a call after to return completed true. I think that would help add confidence in the current implementation if we can see that the second call to Convert may be required.

Reply:
Could we put a reasonable upper bound (rather than a while true)? Like will it take more than 10 iterations?